### PR TITLE
feat: add w3c headers to http request

### PR DESCRIPTION
**Pull request by saartochner**

### CODE REVIEW

**Review by nirLumigo — APPROVED** [review]

File: src/utils/w3cUtils.test.ts [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+    spies.getCurrentTransactionId.mockReturnValueOnce('111111111111112222222222');
+
+    addW3CTracePropagator(headers);
+
+    const parts = headers[TRACEPARENT_HEADER_NAME].split('-');
+    expect(parts.length).toEqual(4);
+    expect(parts[0]).toEqual('00');
+    expect(parts[1]).toEqual('11111111111111111111111100000000');
+    expect(parts[3]).toEqual('01');
+    expect(headers[TRACESTATE_HEADER_NAME]).toEqual(`old,lumigo=${parts[2]}`);
```
Comment (nirLumigo) [inline src/utils/w3cUtils.test.ts:29]: Why not just `==`?

File: src/utils/w3cUtils.ts [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+  headers[TRACESTATE_HEADER_NAME] = getTraceState(headers, messageId);
+  return headers;
+};
+
+const getTraceId = (
+  headers: Record<string, string>,
+  transactionId: string,
+  messageID: string
+): string => {
+  let version = null;
```
Comment (nirLumigo) [inline src/utils/w3cUtils.ts:26]: Can you add comments that explaining this function?

**Review by nirLumigo — APPROVED** [review]

File: src/utils/w3cUtils.test.ts (hunk 1 above)
Comment (saartochner) [reply src/utils/w3cUtils.test.ts:29]: we use expect everywhere in the tracer's tests 🤷

**Review by GuyMoses — APPROVED** [review]

File: src/utils/w3cUtils.ts [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+  let trace_id = null;
+  let span_id = null;
+  let trace_flags = null;
+  const existingHeader = headers?.[TRACEPARENT_HEADER_NAME] || '';
+  const match = TRACEPARENT_HEADER_FORMAT.exec(existingHeader);
+  if (match) {
+    version = match[1];
+    trace_id = match[2];
+    span_id = match[3];
+    trace_flags = match[4];
```
Comment (doriaviram) [inline src/utils/w3cUtils.ts:38]: why you using snake case ?

Comment (doriaviram) [inline src/utils/w3cUtils.ts:38]:
```suggestion
    const [version, traceId, xx] = match
    version = match[1];
    trace_id = match[2];
    span_id = match[3];
    trace_flags = match[4];
```

File: src/utils/w3cUtils.ts [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+    return headers[TRACESTATE_HEADER_NAME] + ',' + lumigoState;
+  }
+  return lumigoState;
+};
+
+export const isW3CHeaders = (headers: Record<string, string>): boolean =>
+  !!TRACEPARENT_HEADER_FORMAT.exec(headers?.[TRACEPARENT_HEADER_NAME] || '');
+
+export const getW3CMessageId = (headers: Record<string, string>): string | null => {
+  const match = TRACEPARENT_HEADER_FORMAT.exec(headers?.[TRACEPARENT_HEADER_NAME] || '');
```
Comment (doriaviram) [inline src/utils/w3cUtils.ts:66]: You run this regex over and over again, this part is not readable. it's better to create a function that extracts `traceId` + `spanId` + `versions` from header. 
and then in this part the reader should understand you want to take `traceId` into `messageId` and not total guess what is match[3]`

File: src/parsers/aws.ts [hunk 5]
```diff
@@ -178,13 +179,17 @@ export const kinesisParser = (requestData, responseData) => {
...
+export const defaultParser = (requestData, responseData) => {
   if (!responseData) return {};
   const { headers: resHeader } = responseData;
-  const messageId = resHeader
+  let messageId = resHeader
     ? resHeader['x-amzn-requestid'] || resHeader['x-amz-request-id']
     : undefined;
 
+  if (!messageId && isW3CHeaders(requestData.headers)) {
+    messageId = getW3CMessageId(requestData.headers);
```
Comment (doriaviram) [inline src/parsers/aws.ts:190]: Don't you run the regex twice here?

**Pull request merged**